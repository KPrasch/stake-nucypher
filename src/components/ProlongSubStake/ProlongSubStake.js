import React, { useState } from 'react';
import { Form, Col, Row, Button } from 'react-bootstrap';
import { Formik } from 'formik';
import Loading from '../Loading/Loading';
import * as Yup from 'yup';

import './ProlongSubStake.scss';

const validationSchema = Yup.object().shape({
  prolongDurationValue: Yup.number().min(1, 'Minimum prolongation is 1 period').required('Field is required.')
});

function ProlongSubStake(props) {
  const onSubStakeProlong = (stakeValues) => {
    if (props.onSubStakeProlong) {
      props.onSubStakeProlong(stakeValues);
    }
  };
  const [loading, setLoading] = useState(false);
  const onSubmit = async (...args) => {
    setLoading(true);
    if (onSubStakeProlong) {
      await onSubStakeProlong(...args);
    }
    setLoading(false);
  };
  return <Formik onSubmit={onSubmit} initialValues={{ prolongDurationValue: 1 }} validationSchema={validationSchema}>
    {({
      handleSubmit,
      handleChange,
      handleBlur,
      values,
      touched,
      isValid,
      errors,
      setValues,
      setErrors
    }) => {
      if (!loading) {
        return <Form onSubmit={handleSubmit}>
          <Form.Group as={Row}>
            <Form.Label column md={{ span: 2, offset: 2 }} htmlFor="substake-prolong-duration">
              Extend for
            </Form.Label>
            <Col md={4}>
              <Form.Control id="substake-prolong-duration" type="number" name="prolongDurationValue" onChange={handleChange} value={values.prolongDurationValue} />
              <div className="feedback-placeholder">
                <Form.Control.Feedback type="invalid" style={{ display: errors.prolongDurationValue && touched.prolongDurationValue ? 'inline' : 'none' }}>{errors.prolongDuration}</Form.Control.Feedback>
              </div>
            </Col>
            <Form.Label column md={4}>
              periods
            </Form.Label>
          </Form.Group>
          <div className="d-flex justify-content-center">
            <Button type="submit"  disabled={!isValid}>Prolong</Button>
          </div>
        </Form>;
      } else {
        return <>
          <div className="mt-2 d-flex justify-content-center align-items-center">
            <Loading size={60}></Loading>
          </div>
        </>;
      }
    }
  }
  </Formik>;
}

export default ProlongSubStake;
