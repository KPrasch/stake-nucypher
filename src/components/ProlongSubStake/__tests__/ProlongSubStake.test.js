import React from 'react';
import ProlongSubStake from '../ProlongSubStake';
import { render, fireEvent, wait } from '@testing-library/react'

describe('<ProlongSubStake />', () => {

  it('should render correctly', () => {
    const { asFragment } = render(<ProlongSubStake />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should trigger onSubStakeProlong on prolongation', async () => {
    const onSubStakeProlong = jest.fn(() => {});
    const { getByLabelText, getByText } = render(<ProlongSubStake onSubStakeProlong={onSubStakeProlong} />);

    await wait(() => {
      fireEvent.change(getByLabelText('Extend for'), { target: { value: '21' } });
      fireEvent.click(getByText('Prolong'));
    });

    expect(onSubStakeProlong).toBeCalledWith(expect.objectContaining({ prolongDurationValue: 21 }));
  });

  it('should validate input from user correctly', async () => {
    const { getByLabelText, getByText, asFragment } = render(<ProlongSubStake />);

    await wait(() => {
      fireEvent.change(getByLabelText('Extend for'), { target: { value: '0' } });
      fireEvent.click(getByText('Prolong'));
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
