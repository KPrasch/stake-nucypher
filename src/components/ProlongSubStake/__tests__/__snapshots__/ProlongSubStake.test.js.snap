// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProlongSubStake /> should render correctly 1`] = `
<DocumentFragment>
  <form
    class=""
  >
    <div
      class="form-group row"
    >
      <label
        class="form-label col-form-label col-md-2 offset-md-2"
        for="substake-prolong-duration"
      >
        Extend for
      </label>
      <div
        class="col-md-4"
      >
        <input
          class="form-control"
          id="substake-prolong-duration"
          name="prolongDurationValue"
          type="number"
          value="1"
        />
        <div
          class="feedback-placeholder"
        >
          <div
            class="invalid-feedback"
            style="display: none;"
          />
        </div>
      </div>
      <label
        class="form-label col-form-label col-md-4"
      >
        periods
      </label>
    </div>
    <div
      class="d-flex justify-content-center"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Prolong
      </button>
    </div>
  </form>
</DocumentFragment>
`;

exports[`<ProlongSubStake /> should validate input from user correctly 1`] = `
<DocumentFragment>
  <form
    class=""
  >
    <div
      class="form-group row"
    >
      <label
        class="form-label col-form-label col-md-2 offset-md-2"
        for="substake-prolong-duration"
      >
        Extend for
      </label>
      <div
        class="col-md-4"
      >
        <input
          class="form-control"
          id="substake-prolong-duration"
          name="prolongDurationValue"
          type="number"
          value="0"
        />
        <div
          class="feedback-placeholder"
        >
          <div
            class="invalid-feedback"
            style="display: inline;"
          />
        </div>
      </div>
      <label
        class="form-label col-form-label col-md-4"
      >
        periods
      </label>
    </div>
    <div
      class="d-flex justify-content-center"
    >
      <button
        class="btn btn-primary"
        disabled=""
        type="submit"
      >
        Prolong
      </button>
    </div>
  </form>
</DocumentFragment>
`;
